### PR TITLE
Development for StokesFOHydrology problem

### PR DESCRIPTION
**Pull request by bartgol**
This PR introduces all the development I did for the FOHydrology problem. It is still a work in progress, but I wanted to put it up, since it's a huge PR, and wanted to give ppl a heads up that there are big changes coming up.

Here's a brief summary of what's in here:

- Remove some evaluators: PHAL::SideToCell (never used), LandIce::IceSoftness (same as FlowRate), LandIce::ProlongateVector (never used, I think I added it years ago, when I was thinking about coupling 3d and 2d), and replaced PHAL::QuadPointsToCellInterpolation, PHAL::SideQuadPointsToSideInterpolation, and PHAL::NodesToCellInterpolation with a single PHAL::P0Interpolation.
- Removed SchoofFit problem: I used it once, roughly 5 years ago, and never again.
- Added PHAL::GatherSolutionSide evaluator (better than doing GatherSolution+CellToSide).
- All hydrology evaluators are in src/LandIce/evaluators/hydrology
- Removed some template args from some LandIce evaluators: if a template arg is needed to deduce the scalar type of a field, then it is needed, otherwise it can be traded for a runtime switch, saving compilation time.
- Miscellanea improvement to StokesFOBase "generic" (i.e., interpolation-like) evaluators construction. Notably, use PHX to figure out if, say, a QuadPoint version of field F is available already (perhaps computed by some physics ev); if not, build interpolation ev, otherwise skip. Note: this requires the interpolation evaluators to be built *last* in the `constructEvaluators` call.
- Add FOHydrology problem: pretty much what you'd expect by merging StokesFOBase with Hydrology implementations.
- More support for collapsed sideset layouts. Notably, in FOThermoCoupled.

As of today, all tests pass, except for the new FOHydrology one, and FO_GIS_CoupledThickness (both Epetra and Tpetra). The latter runs, but fails the comparison at the end. I looked at the output, and in the branch there are small differences in the residual, which end up triggering an extra NOX iteration, which yield a different response. I am not sure where the diffs come from, I am trying to figure this out.

### CODE REVIEW

**Review by mperego — COMMENTED** [review]
Lots of changes!
It looks good to me, when you are ready to merge let me know and I'll test it with MPAS and other few things.

**Comment by mperego** [comment]
@bartgol does the gatherSolutionSide need to be Kokkosified?

File: src/LandIce/evaluators/LandIce_IceOverburden_Def.hpp.orig
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_IceOverburden_Def.hpp.orig:1]: Ah, good call.

Comment (bartgol) [reply src/LandIce/evaluators/LandIce_IceOverburden_Def.hpp.orig:1]: Uhm, for some reason it is still used. I'm wondering if something got messed up with the rebase. But yeah, the idea is to remove it.

File: src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp:240]: Ah, that might fix the test failing!

File: src/LandIce/evaluators/hydrology/LandIce_EffectivePressure_Def.hpp.orig
Comment (bartgol) [reply src/LandIce/evaluators/hydrology/LandIce_EffectivePressure_Def.hpp.orig]: Good catch! I should always run `find . *.orig` after a rebase/merge...

**Comment by bartgol** [comment]
> @bartgol does the gatherSolutionSide need to be Kokkosified?

Yes, I haven't Kokkosified it. But it's not used by current LandIce problems (there was a use in StokesFO, but it is commented out), so it should not cause tests to fail on GPU.

File: src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp:240]: It did! I'm a bit puzzled: how come the buggy line didn't cause OOB reads? I am assuming PHX (or Kokkos) would turn on OOB access in debug builds in trilinos, but maybe I'm wrong. But even if PHX/Kokkos don't check, it's hard to believe that the data read out of bounds was good enough to a) not cause NaN's, and b) allow convergence.

Comment (bartgol) [reply src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp:240]: Nvm, it makes sense. The sideSetView's 1st dim is `worksetSize*numSides`, so using the element LID will never put you out of bounds.

**Comment by bartgol** [comment]
@mperego All (existing) tests pass on my laptop now. I don't think we should wait on the FO Hydro test to pass to merge this.

If you want to test MPAS and other stuff, I think now it's ok to do so.

**Comment by mperego** [comment]
> 
> 
> @mperego All (existing) tests pass on my laptop now. I don't think we should wait on the FO Hydro test to pass to merge this.
> 
> If you want to test MPAS and other stuff, I think now it's ok to do so.

OK, will do.

**Comment by bartgol** [comment]
I added a cmake option `ENABLE_EXPERIMENTAL`, which we can use to push to master code/tests that we are currently developing, without having to comment out codes/includes. It seemed like a neater option.

**Comment by bartgol** [comment]
@mcarlson801 I added you as a reviewer so you can check if I messed up stuff related to collapsed sidesets. Could you please take a look at that? Sorry for not giving you a heads up; I understand if you don't have time.

**Comment by mperego** [comment]
@bartgol where you planning to move the `LandIce_HydrostaticPressure` evaluator in the `LandIce/evaluators/hydrology` folder? In the `src/LandIce/CMakeLists.txt` you did so, but the actual files are still in the `LandIce/evaluators`

**Comment by bartgol** [comment]
> @bartgol were you planning to move the `LandIce_HydrostaticPressure` evaluator in the `LandIce/evaluators/hydrology` folder? In the `src/LandIce/CMakeLists.txt` you did so, but the actual files are still in the `LandIce/evaluators`

Ah, no,  I wasn't. We don't use that evaluator in the Hydrology problem(s).

On this note, I think there's some overlap between hydrology evaluators and enthalpy ones (e.g. basal melt rate). There might be some differences, but perhaps we can unify the duplicates into one, and handle minor model differences internally in the evaluator calls. That would reduce the list of evaluators to compile. Not in this PR though.

**Comment by bartgol** [comment]
Ah, I see why everything compiled: the cpp file is still listed correctly in the CMakeLists.txt, it's only the header files that are listed in the wrong folder. I should run make install to check everything is ok with headers.

**Comment by mperego** [comment]
> 
> 
> Ah, I see why everything compiled: the cpp file is still listed correctly in the CMakeLists.txt, it's only the header files that are listed in the wrong folder. I should run make install to check everything is ok with headers.

yeah, I noticed because I tried to install Albany.

**Comment by mperego** [comment]
@bartgol there are some tests failing when compiled with the "MESH_DEPENDS_ON_PARAMETERS" flag

**Comment by bartgol** [comment]
> @bartgol there are some tests failing when compiled with the "MESH_DEPENDS_ON_PARAMETERS" flag

Ah, crap, I didn't test those either. Ok, I will run again trying to use all the capabilities.

**Comment by mperego** [comment]
@bartgol  there is also a FROSch test that fails. You do not need to install FROSch, you can use Ifpack2 instead. Here is the file of the test failing:
`https://github.com/SNLComputation/Albany/blob/master/tests/small/LandIce/FO_GIS/input_fo_humboldt_frosch.yaml`

**Comment by bartgol** [comment]
> @bartgol there is also a FROSch test that fails. You do not need to install FROSch, you can use Ifpack2 instead. Here is the file of the test failing:
> `https://github.com/SNLComputation/Albany/blob/master/tests/small/LandIce/FO_GIS/input_fo_humboldt_frosch.yaml`

Ah, I see I need to enable ALBANY_FROSH to get the test enabled. But that's enabled internally if FROSch is found. I will just install frosch in trilinos, so I won't miss this test later.

File: src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp
Comment (mcarlson801) [inline src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp:238]: This comment isn't specifically related to this line or even this PR but it is the first time I've taken a look at this evaluator and I noticed something that may cause issues in the future.

It looks like for every single value of iSide, a number of dynamic rank views are being allocated. This is fine for CPU I imagine, but this will need to be changed eventually for GPU since allocating many small arrays on the device becomes much more expensive. Again, it isn't relevant for this PR but it might be worth adding a TODO comment.

File: src/LandIce/evaluators/LandIce_w_Resid_Def.hpp
Comment (mcarlson801) [inline src/LandIce/evaluators/LandIce_w_Resid_Def.hpp:55]: sideBF, side_w_measure, side_w_qp, basalVerticalVelocitySideQP, and normals need to set layout to the _sideset variant if useCollapsedSidesets is set. Looks like it is still using the non-collapsed layout.

File: src/LandIce/evaluators/LandIce_w_Resid_Def.hpp
Comment (mcarlson801) [inline src/LandIce/evaluators/LandIce_w_Resid_Def.hpp:132]: Since there is only an implementation for the collapsed layout case, is it safe to assume that the old layouts will never be used for this evaluator? If so, we can probably remove the useCollapsedLayout checks in this evaluator and just use the new layout. 

The useCollapsedLayout stuff is only necessary for evaluators that may be used by multiple problems where one of the problems hasn't been updated yet. I think this one is only used by the coupled problem, right?

**Review by mcarlson801 — CHANGES_REQUESTED** [review]
Looks good! I've left a few comments relating to the sideset layouts. Most are just minor comments and thoughts about the future but there is one relating to w_resid that will probably result in sneaky crashes and should be fixed before going live.

File: src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_ThicknessResid_Def.hpp:238]: Yes, I think you're right. BTW, I think this was somewhat copied from what happens in Neumann evaluator. Except in Neumann, the temporaries are created in postRegSetup, with a size large enough to accommodate all sides, and during evaluation, they are subviewed at the proper offset location. Maybe the same mod can be done here too. Or maybe we should just go with raw Kokkos views (I don't understand the need for DynRankView).

File: src/LandIce/evaluators/LandIce_w_Resid_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_w_Resid_Def.hpp:55]: I think you might have missed the lines right above: to avoid long lines in the construction of the PHX fields, I created the layouts above, with a switch on `useCollapsedSidesets`. Does that make sense?

Comment (mcarlson801) [reply src/LandIce/evaluators/LandIce_w_Resid_Def.hpp:55]: Oh woops, I see that now. I looked at this too quickly and misunderstood. Looks fine as is.

File: src/LandIce/evaluators/LandIce_w_Resid_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_w_Resid_Def.hpp:132]: Good point. The impl is only for collapsed, so attempting to create the ev with non-collapsed layouts might set you up for a nice ride...

File: src/LandIce/evaluators/LandIce_w_Resid_Def.hpp
Comment (bartgol) [reply src/LandIce/evaluators/LandIce_w_Resid_Def.hpp:52]: Ah right. I always mix up asserts and teuchos' exceptions.

**Review by mcarlson801 — APPROVED** [review]

**Comment by bartgol** [comment]
I finally fixed all the FO tests, and I'm down to only the SHMIP ones failing. This is for MESH_DEPENDS_ON_PARAMETERS=ON, the other case already worked. I am planning to finish off the SHMIP tests tonight or tomorrow morning. Then I will re-test everything (hoping to have not screwed up stuff in other tests). Fingers crossed, this PR might be ready to merge tomorrow.

**Comment by bartgol** [comment]
The MESH_DEPENDS_ON_PARAMETER build passes all tests on my laptop. I'm re-running the "normal" build again, to be sure it didn't screw up anything. If there are no objections, and tests pass, I'll merge by COB on Tuesday.

**Comment by bartgol** [comment]
I fixed the last test. Everything passes on my machine, with MESH_DEPENDS_ON_XYZ both ON and OFF. Per our discussion at the weekly telecon, I'm merging this.

**Pull request merged**
